x/dsmr/node_test.go

```diff
@@ -994,11 +1057,220 @@ func TestNode_NewBlock_IncludesChunkCerts(t *testing.T) {
+			parentBlocks:  [][]int{{0, 2}},
+			chunks:        []int{2, 4},
+			timestamp:     3,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  nil, // build would filter out duplicate chunks, hence no error.
+		},
+		{
+			name:          "one block duplicate chunks",
+			parentBlocks:  [][]int{},
+			chunks:        []int{2, 2},
+			timestamp:     2,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  nil, // build would filter out duplicate chunks, hence no error.
+		},
+		{
+			name:          "three blocks non consecutive duplicate chunks",
+			parentBlocks:  [][]int{{3}, {2}},
+			chunks:        []int{3},
+			timestamp:     4,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  ErrNoAvailableChunkCerts,
+		},
+		{
+			name:          "monotonic timestamping",
+			parentBlocks:  [][]int{},
+			chunks:        []int{1},
+			timestamp:     0,
+			verifyWantErr: ErrInvalidBlockHeight,
+			buildWantErr:  ErrTimestampNotMonotonicallyIncreasing,
+		},
+		{
+			name:          "empty block",
+			parentBlocks:  [][]int{{1}, {2}, {3}, {4}},
+			chunks:        []int{},
+			timestamp:     6,
+			verifyWantErr: ErrEmptyBlock,
+			buildWantErr:  ErrNoAvailableChunkCerts,
+		},
+		{
+			name:                       "three blocks non consecutive duplicate chunks outside validity window",
+			parentBlocks:               [][]int{{4}, {5}},
+			chunks:                     []int{4},
+			timestamp:                  4,
+			verifyWantErr:              validitywindow.ErrDuplicateContainer, // this isn't ideal, since it would disqualify a duplicate chunk even when it's outside the validity window. However, it default to the correct direction.
+			buildWantErr:               ErrNoAvailableChunkCerts,
+			validalidityWindowDuration: 1,
+		},
+	}
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			r := require.New(t)
+			validationWindow := testCase.validalidityWindowDuration
+			if validationWindow == 0 {
+				validationWindow = testingDefaultValidityWindowDuration
+			}
+
+			var node *Node[tx]
+			nodes, indexer := newNodes(t, 1)
+			node = nodes[0]
+			node.validityWindowDuration = validationWindow
+
+			// initialize node history.
+			parentBlk := node.LastAccepted
+			for _, chunkList := range testCase.parentBlocks {
+				blk := Block{
+					ParentID:  parentBlk.GetID(),
+					Height:    uint64(int(node.LastAccepted.Height) + 1),
+					Timestamp: int64(int(node.LastAccepted.Timestamp) + 1),
+					blkID:     ids.GenerateTestID(),
+				}
+
+				for _, chunkExpiry := range chunkList {
+					_, chunkCert, err := node.BuildChunk(
+						context.Background(),
+						[]tx{
+							{
+								ID:     ids.Empty.Prefix(uint64(chunkExpiry)),
+								Expiry: int64(chunkExpiry),
+							},
+						},
+						int64(chunkExpiry),
+						codec.Address{},
+					)
+					r.NoError(err)
+					blk.ChunkCerts = append(blk.ChunkCerts, &chunkCert)
+				}
```

Comment:
Can we use `BuildBlock` after calling `BuildChunk` instead of initializing the block through the struct?